### ISSUE TITLE
--help output embeds randomly generated default, making docs non-reproducible

### ISSUE THREAD

**Issue by paravoid** [tox-dev/tox#2942]
## Issue

The `tox legacy --help` output changes on every invocation because of the `--hashseed` default:
```python
parser.add_argument(
    "--hashseed",
    metavar="SEED",
    help="set PYTHONHASHSEED to SEED before running commands. Defaults to a random integer in the range "
    "[1, 4294967295] ([1, 1024] on Windows). Passing 'noset' suppresses this behavior.",    
    action=SeedAction,
    of_type=Optional[int],
    default=random.randint(1, 1024 if sys.platform == "win32" else 4294967295),
    dest="hash_seed",
)
```

Besides the lack of usefulness (or lie, depending on your viewpoint ;) on what the default for this parameter is, the real-world impact of this is that this gets included in the generated Sphinx documentation (by the way of argparse-cli), which makes the build non-reproducible: multiple runs of documentation building generate slightly different output.

## Environment

This is with tox 4.4.6, but I've verified the same applies to current master. This is on a Debian sid system, with (my) Debian packages of tox, but I don't think that makes a difference here.

## Minimal example

```console
$ tox legacy --help | sed -nr 's/(.+--hashseed SEED).+\. (.+)/\1 ... \2/p'
  --hashseed SEED ... (default: 499015295)
$ tox legacy --help | sed -nr 's/(.+--hashseed SEED).+\. (.+)/\1 ... \2/p'
  --hashseed SEED ... (default: 1099173378)
```

**Comment by paravoid** [comment]
I can and that would fix my use case.

My thinking for filing this as an upstream bug though was that this the help output is somewhat misleading: the default isn't "1099173378", it's "<some random number>". Not sure if anyone is not going to be confused in practice though.

**Comment by gaborbernat** [comment]
But that's the default for that run 🤔 so it is accurate, it just changes per run. Not sure what would be a better solution.

**Comment by paravoid** [comment]
Maybe I'm misunderstanding something... I expect what I see in the `--help` output to be useful for the subsequent "run" :) Basically the `--help` output is not a "run", it's just help output.

Or to put it another way: https://tox.wiki/en/latest/cli_interface.html right now says: "default: 17441603". That's not the default! The default is a random value (as the same line says earlier).

I think a solution could be to omit the _"(default: NNN)"_ part from the output, keeping the _"Defaults to a random integer in the range"_ part. Perhaps by removing the default from `parser.add_argument()`, and then having an `if args.hashseed is None: ...` after argument parsing? Do you think that's too minor to be worth it?

**Comment by paravoid** [comment]
Something like (untested):
```diff
     def _default_set_env(self) -> dict[str, str]:
         env = super()._default_set_env()
-        hash_seed: int | None = getattr(self.options, "hash_seed", None)
-        if hash_seed is not None:
-            env["PYTHONHASHSEED"] = str(hash_seed)
+        default_seed = random.randint(1, 1024 if sys.platform == "win32" else 4294967295)
+        hash_seed: int = getattr(self.options, "hash_seed", default_seed)
+        env["PYTHONHASHSEED"] = str(hash_seed)
         return env
```

Alternatively:
```diff
     def _default_set_env(self) -> dict[str, str]:
         env = super()._default_set_env()
-        hash_seed: int | None = getattr(self.options, "hash_seed", None)
-        if hash_seed is not None:
-            env["PYTHONHASHSEED"] = str(hash_seed)
+        env["PYTHONHASHSEED"] = str(getattr(self.options, "hash_seed", "random")
         return env
```
(...as PYTHONHASHSEED actually accepts the string literal "random")

**Comment by gaborbernat** [comment]
> I expect what I see in the `--help` output to be useful for the subsequent "run"

I think this is the disconnect here. I do not expect this to hold true. 



> . Perhaps by removing the default from `parser.add_argument()`, and then having an `if args.hashseed is None: ...` after argument parsing?

IMHO what we have is correct, there's no reason to have this disconnect of post processing, while we don't do this for any other value or config.

At best I could see an argument to mask for the user the hash_seed as `<random>`.

**Comment by paravoid** [comment]
> > I expect what I see in the `--help` output to be useful for the subsequent "run"
> 
> I think this is the disconnect here. I do not expect this to hold true.

OK. Do tell!

> > . Perhaps by removing the default from `parser.add_argument()`, and then having an `if args.hashseed is None: ...` after argument parsing?
> 
> IMHO what we have is correct, there's no reason to have this disconnect of post processing, while we don't do this for any other value or config.
> 
> At best I could see an argument to mask for the user the hash_seed as `<random>`.

Taking a step or two back... two questions:
- PYTHONHASHSEED accepts the string literal `"random"`, and performs the "pick random integer" internally in the C code. I guess you were aware and had your reasons to reimplement that in tox with random.int()?
- Shouldn't tox inherit PYTHONHASHSEED from its own environment? So if I run `PYTHONHASHSEED=5 tox ...`, shouldn't tox use that value? That would perhaps help, because we could perhaps do `PYTHONHASHSEED=0 sphinx-build ...` but sounds like a good idea regardless?

**Comment by gaborbernat** [comment]
> * PYTHONHASHSEED accepts the string literal `"random"`, and performs the "pick random integer" internally in the C code. I guess you were aware and had your reasons to reimplement that in tox with random.int()?

This would randomize the seed for each invocation. I want to have same randmome for each call within a single run, and also the seed should be available for plugins.

> * Shouldn't tox inherit PYTHONHASHSEED from its own environment? So if I run `PYTHONHASHSEED=5 tox ...`, shouldn't tox use that value? That would perhaps help, because we could perhaps do `PYTHONHASHSEED=0 sphinx-build ...` but sounds like a good idea regardless?

We do not support this kind of invoke documentation build on your own behaviour, and was a concious decision. People should use `tox` to generate the documentation, within distributions probably through something like https://github.com/fedora-python/tox-current-env.

As for the tox supporting PYTHONHASHSEED env var I'm open to that, though you can do that today too with `tox run -e doc --hashseed 5` 🤔 so no need for direct support via env-var. I prefer the explicitness of the CLI flag over the hidden nature of env vars here.

**Comment by paravoid** [comment]
> > * PYTHONHASHSEED accepts the string literal `"random"`, and performs the "pick random integer" internally in the C code. I guess you were aware and had your reasons to reimplement that in tox with random.int()?
> 
> This would randomize the seed for each invocation. I want to have same randmom for each call within a single run, and also the seed should be available for plugins.

Oh I see. That makes total sense.

> > * Shouldn't tox inherit PYTHONHASHSEED from its own environment? So if I run `PYTHONHASHSEED=5 tox ...`, shouldn't tox use that value? That would perhaps help, because we could perhaps do `PYTHONHASHSEED=0 sphinx-build ...` but sounds like a good idea regardless?
> 
> We do not support this kind of invoke documentation build on your own behaviour, and was a concious decision. People should use `tox` to generate the documentation, within distributions probably through something like https://github.com/fedora-python/tox-current-env.

I didn't know about tox-current-env. That's a useful tip, I'll look more into it. However...

> As for the tox supporting PYTHONHASHSEED env var I'm open to that, though you can do that today too with `tox run -e doc --hashseed 5` thinking so no need for direct support via env-var. I prefer the explicitness of the CLI flag over the hidden nature of env vars here.

Again, I may be misunderstanding something, but as I understand it that won't cut it:
- `tox -e docs --hashseed 5` sets `os.environ['PYTHONHASHSEED']`, then calls `sphinx-build` (as outlined on tox.ini)
- sphinx-build (by the way of argparse-cli etc.) imports the tox source code and runs `_get_parser_doc()`, `_get_parser()`, `tox_add_option()` etc.
- `parser.add_argument()` is called to generate its text output, but its `default` does not obey PYTHONHASHSEED (i.e. what I proposed above) -- it just generates a new random number every time.
-  The generated documentation includes a randomly picked integer every time.

In other words: `tox run -e docs --hashseed 5; grep tox-legacy---hashseed .tox/docs_out/cli_interface.html` will still be different after every run.

Side-notes:
-  `tox -e docs` doesn't really generate docs in current master as far as I can tell, because tox calls `sphinx-build ... -b html -b linkcheck ...` and sphinx-build only obeys the last `-b`. So I had to remove ` {posargs:-b linkcheck -W}` from tox.ini to make it work.
- Unrelated to docs-building, but the test suite fails with `tox --hashseed 5`, due to some difference in Requirement ordering  in the `test_load_dependency_many_extra` & `test_load_dependency_no_extra` tests.

**Comment by paravoid** [comment]
> > In other words: `tox run -e docs --hashseed 5; grep tox-legacy---hashseed .tox/docs_out/cli_interface.html` will still be different after every run.
> 
> It would be the same random, though no? Because if sphinx is called with the same seed it will generate the same random number every time, no?

My understanding is that PYTHONHASHSEED only affects the seed for `hash()`, and does not affect `random.randint()` etc. (i.e. does not call `random.seed()` internally). So, no, it would not be the same random.

In other words, repeated invocations of `PYTHONHASHSEED=5 python3 -c 'import random; print(random.randint(1, 1024))'` on a shell results into different numbers every time.

**Comment by paravoid** [comment]
Do you mean something like this?

```python
--- a/src/tox/session/cmd/run/common.py
+++ b/src/tox/session/cmd/run/common.py
@@ -131,6 +131,14 @@ def env_run_create_flags(parser: ArgumentParser, mode: str) -> None:
                         raise ArgumentError(self, str(exc))
                 setattr(namespace, self.dest, result)
 
+        try:
+            hashseed_default = int(os.environ["PYTHONHASHSEED"])
+        except (KeyError, ValueError):
+            if sys.platform == "win32":
+                hashseed_default = random.randint(1, 1024)
+            else:
+                hashseed_default = random.randint(1, 4294967295)
+
         parser.add_argument(
             "--hashseed",
             metavar="SEED",
@@ -138,7 +146,7 @@ def env_run_create_flags(parser: ArgumentParser, mode: str) -> None:
             "[1, 4294967295] ([1, 1024] on Windows). Passing 'noset' suppresses this behavior.",
             action=SeedAction,
             of_type=Optional[int],
-            default=random.randint(1, 1024 if sys.platform == "win32" else 4294967295),
+            default=hashseed_default,
             dest="hash_seed",
         )
     parser.add_argument(
```

**Comment by gaborbernat** [comment]
Minus the try/catch and making the catch part two lines 😂

**Comment by paravoid** [comment]
> Minus the try/catch and making the catch part two lines

I think I understand the second part, but don't think I understand what you mean with the first part. Happy to take your feedback and adjust, but perhaps it'd be easier if you just did it the way you wanted to?

**Comment by gaborbernat** [comment]
Don't use try/catch, instead use if/else 😊
